### DISCUSSION TITLE
Does connectorx support AD auth to MSSQL from linux?

### DISCUSSION THREAD

**Discussion by cawoodjm** [sfu-db/connector-x#242] [Q&A]
Does connectorx ("==0.2.4a9") support connections to MSSQL via AD auth. If I was using sqlalchemy I will use

`sqlalchemy.create_engine(f'mssql+pymssql://{ms_sql_user}:{ms_sql_password}@{ms_sql_server}/{ms_sql_db}')`

, but when I add mssql+pymssql to the string passed to connectorx I get a "not supported" error. Side note, these connections are being made from linux. Typically to accomplish AD auth fromnon-windows systems a third party tool must be used, namely FreeTDS. Which in our case Free TDS is packaged inside the python module pymssql.

Thoughts on the matter?

**Comment by wangxiaoying** [comment]
Hi @cawoodjm , we support [integrated authentication](https://docs.rs/tiberius/latest/tiberius/enum.AuthMethod.html#variant.Integrated) for mssql. To enable it, the connection string looks like this: `mssql://host:port/db?trusted_connection=true`. But we haven't tested on linux before, can you try out?

**Reply by cawoodjm** [reply]
I stumbled upon the `?trusted_connection=true` last night and it seems too timeout when trying to connect.
`ERROR tiberius::tds::stream::token] message=Login failed for user ''.` I will note, I tested on Linux with a server account using connectorx and it works. Not sure if FreeTDS is needed to make it work.

Python Version: 3.8
OS Ubuntu 20.04 LTS
DB: MS Sql Server 2016
connectorx version: 0.2.4a9

**Reply by cawoodjm** [reply]
@wangxiaoying I do not want to place all the weight on the connectorx team to develop trusted connection capability on linux hardware, but lack in knowledge to implement such a Free TDS solution in the Ruse framework. I would be willing to help or test where I can. Maybe even point me to any code snippets that would get me down the road for a pull request.

**Reply by wangxiaoying** [reply]
Hi @cawoodjm , thank you! It would be great if you could help us in this since we don't have the environment.

Currently we are using [tiberius](https://docs.rs/tiberius/latest/tiberius/) as the client driver to connect to mssql. And for AD auth we use its [integrated authentication](https://docs.rs/tiberius/latest/tiberius/enum.AuthMethod.html#variant.Integrated). The code related to mssql authentication can be found [here](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mssql/mod.rs#L46). Actually we don't support it on linux right now since it requires extra feature `integrated-auth-gssapi` and we cannot test it (sorry that I forgot about it previously). In order to support linux AD auth, I think we need to enable this feature and remove these [os conditions](https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/mssql/mod.rs#L66), which may also require to install some extra dependency like mentioned in this [doc](https://github.com/prisma/tiberius/blob/81222b0d110b2bbb7df70090ca47b69805d73e54/README.md#integrated-authentication-trustedconnection-on-nix). Please let me know if you need anything from my side!
